hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDeletingService.java

```diff
@@ -0,0 +1,368 @@
+                    keyInfo, previousKeyTable);
+              }
+
+              toActiveDBList.add(toActiveDb.build());
+              toNextDBList.add(toNextDb.build());
+
+            }
+            // Submit Move request to OM.
+            submitSnapshotMoveDeletedKeys(snapInfo, toActiveDBList,
+                toNextDBList);
+            snapshotLimit--;
+            successRunCount.incrementAndGet();
+          } catch (IOException ex) {
+            LOG.error("Error while running Snapshot Deleting Service", ex);
+          }
+        }
+      } catch (IOException e) {
+        LOG.error("Error while running Snapshot Deleting Service", e);
+      }
+
+      return BackgroundTaskResult.EmptyTaskResult.newResult();
+    }
+
+    private void splitRepeatedOmKeyInfo(SnapshotMoveKeyInfos.Builder toActiveDb,
+        SnapshotMoveKeyInfos.Builder toNextDb, OmKeyInfo keyInfo,
+        Table<String, OmKeyInfo> previousKeyTable) throws IOException {
+      if (checkKeyExistInPreviousTable(previousKeyTable, keyInfo)) {
+        // Move to next non deleted snapshot's deleted table
+        toNextDb.addKeyInfos(keyInfo.getProtobuf(
+            ClientVersion.CURRENT_VERSION));
+      } else {
+        // Move to active DB Deleted Table.
+        toActiveDb.addKeyInfos(keyInfo
+            .getProtobuf(ClientVersion.CURRENT_VERSION));
+      }
+    }
+
+    private void submitSnapshotMoveDeletedKeys(SnapshotInfo snapInfo,
+        List<SnapshotMoveKeyInfos> toActiveDBList,
+        List<SnapshotMoveKeyInfos> toNextDBList) {
+
+      SnapshotMoveDeletedKeysRequest.Builder moveDeletedKeysBuilder =
+          SnapshotMoveDeletedKeysRequest.newBuilder()
+              .setFromSnapshot(snapInfo.getProtobuf());
+
+      SnapshotMoveDeletedKeysRequest moveDeletedKeys =
+          moveDeletedKeysBuilder.addAllActiveDBKeys(toActiveDBList)
+          .addAllNextDBKeys(toNextDBList).build();
+
+
+      OMRequest omRequest = OMRequest.newBuilder()
+          .setCmdType(Type.SnapshotMoveDeletedKeys)
+          .setSnapshotMoveDeletedKeysRequest(moveDeletedKeys)
+          .setClientId(clientId.toString())
+          .build();
+
+      submitRequest(omRequest);
+    }
+
+    private boolean checkKeyExistInPreviousTable(
+        Table<String, OmKeyInfo> previousKeyTable, OmKeyInfo deletedKeyInfo)
+        throws IOException {
+
+      if (previousKeyTable == null) {
+        return false;
+      }
+
+      //TODO: [SNAPSHOT] Handle Renamed Keys
+      String dbKey = ozoneManager.getMetadataManager()
+          .getOzoneKey(deletedKeyInfo.getVolumeName(),
+              deletedKeyInfo.getBucketName(), deletedKeyInfo.getKeyName());
+
+      OmKeyInfo prevKeyInfo = previousKeyTable.get(dbKey);
+      if (prevKeyInfo != null &&
+          prevKeyInfo.getObjectID() == deletedKeyInfo.getObjectID()) {
+        return true;
+      }
+      return false;
+    }
+
+    private SnapshotInfo getPreviousSnapshot(SnapshotInfo snapInfo)
```

Comment:
Should this method only return active snapshots?

Reply:
No it doesn't. I should return the previous Snapshot regarless.